Firestore/core/src/local/leveldb_key.h

```diff
@@ -874,6 +879,51 @@ class LevelDbIndexEntryKey {
   std::string document_key_;
 };
 
+/** A key in the document_overlays table. */
+class LevelDbDocumentOverlayKey {
+ public:
+  /**
+   * Creates a key prefix that points just before the first key in the table.
+   */
+  static std::string KeyPrefix();
+
+  /**
+   * Creates a key prefix that points just before the first key for the given
+   * user_id.
+   */
+  static std::string KeyPrefix(absl::string_view user_id);
+
+  /** Creates a complete key that points to a specific user_id and document key.
+   */
+  static std::string Key(absl::string_view user_id,
+                         const model::DocumentKey& document_key);
+
+  /**
+   * Decodes the given complete key, storing the decoded values in this
+   * instance.
+   *
+   * @return true if the key successfully decoded, false otherwise. If false is
+   * returned, this instance is in an undefined state until the next call to
+   * `Decode()`.
+   */
+  ABSL_MUST_USE_RESULT
+  bool Decode(absl::string_view key);
+
+  /** The user ID, as encoded in the key. */
+  const std::string& user_id() const {
+    return user_id_;
+  }
+
+  /** The path to the document, as encoded in the key. */
+  const model::DocumentKey& document_key() const {
+    return document_key_;
+  }
+
+ private:
+  std::string user_id_;
+  model::DocumentKey document_key_;
```

Comment:
Just want to point this out before I review further, I think if you add `largest_batch_id` as the third segment to the key, you can avoid the new proto and remove the serialization code as well.

This PR will be much smaller this way.

Reply:
Hmmm okay I'll give that a try.

Reply:
Done. PTAL.

Reply:
Actually... I found a bug. I'm working on a fix.

Reply:
Fixed by cf04e6740bd371738cd62bed71498838a259e3db